File: src/main/java/com/github/alexandrecarlton/idea/settings/layout/configurations/common/before_launch/BeforeLaunchConfigurationSettings.java
package com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonSubTypes.Type;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import com.fasterxml.jackson.annotation.JsonTypeInfo.As;
import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;

@JsonTypeInfo(
  use = Id.NAME,
  include = As.WRAPPER_OBJECT)
@JsonSubTypes({
  @Type(name = "build", value = BuildConfigurationSettings.class),
  @Type(name = "runAnotherConfiguration", value = RunAnotherConfigurationSettings.class),
  @Type(name = "runMavenGoal", value = RunMavenGoalSettings.class)})
public interface BeforeLaunchConfigurationSettings {
}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/layout/configurations/common/before_launch/BeforeLaunchConfigurationSettings.kt
package com.github.alexandrecarlton.idea.settings.layout.configurations.common.before_launch

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeInfo.As
import com.fasterxml.jackson.annotation.JsonTypeInfo.Id
import com.fasterxml.jackson.annotation.JsonTypeName
import java.io.File

@JsonTypeInfo(use = Id.NAME, include = As.WRAPPER_OBJECT)
sealed class BeforeLaunchConfigurationSettings

@JsonTypeName("Build")
object BuildConfigurationSettings : BeforeLaunchConfigurationSettings()

@JsonTypeName("Run Another Configuration")
data class RunAnotherConfigurationSettings(

    @JsonProperty("Name")
    val name: String,

    @JsonProperty("Type")
    val type: RunConfigurationType? = null

) : BeforeLaunchConfigurationSettings()

@JsonTypeName("Run Maven Goal")
data class RunMavenGoalSettings(

    @JsonProperty("Working directory")
    val workingDirectory: File,

    @JsonProperty("Command line")
    val commandLine: String

) : BeforeLaunchConfigurationSettings()

